csharp/src/Microsoft.ML.OnnxRuntime/OrtAllocator.shared.cs

```diff
@@ -310,6 +311,69 @@ protected override bool ReleaseHandle()
         #endregion
     }
 
+    /// <summary>
+    /// This class represents an arbitrary buffer of memory
+    /// allocated and owned by the user. It can be either a CPU, GPU or other device memory
+    /// that can be suitably represented by IntPtr.
+    /// This is just a composite of the buffer related information.
+    /// The memory is assumed to be pinned if necessary and usable immediately
+    /// in the native code.
+    /// </summary>
+    public class OrtExternalAllocation
+    {
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="memInfo">use to accurately describe a piece of memory that this is wrapping</param>
+        /// <param name="shape">shape of this buffer</param>
+        /// <param name="elementType">element type</param>
+        /// <param name="pointer">the actual pointer to memory</param>
+        /// <param name="sizeInBytes">size of the allocation in bytes</param>
+        public OrtExternalAllocation(OrtMemoryInfo memInfo, long[] shape, Tensors.TensorElementType elementType, IntPtr pointer, long sizeInBytes)
```

Comment:
Do we need this to be provided separately to the size that can be inferred from the shape?

Reply:
I was going back and forth on this. Finally, I decided to supply this argument/property to validate that the user knows what he is doing as this a common mistake to not allocate the proper buffer and also for consistency with OrtMemoryAllocation.